### PR TITLE
Search sentiment map's popularity mode

### PR DESCRIPTION
**Pull request by ChenyuHeidiZhang**
Added search interest (popularity) scores for each country, stored parallel to the averageSentiment score.
Added a toggle button for the two modes so that search interest scores are displayed on the map instead of sentiment scores when the toggle button is "checked."

### CODE REVIEW

File: search-sentiments-map/public/scripts/current-data-script.js [hunk 1]
```diff
@@ -35,19 +35,13 @@ function getCurrentCustomSearchData() {
...
-  fetch('/search/' + currentTrend)
-      .then(resultsJsonArray => resultsJsonArray.json()).then(topicResults => {
-        currentCustomSearchData = topicResults;
-        loadCountryData();
-      });
+  fetch('/search/' + trend)
+      .then(resultsJsonArray => resultsJsonArray.json()).then(topicData => {
+    currentCustomSearchData = topicData;
+  }).then(() => {
+    // Reload map with new sentiment or search interest data and relevant coloring.
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/current-data-script.js:44]:
```suggestion
    // Reload map with either new sentiment or search interest data and relevant coloring.
```

File: search-sentiments-map/public/scripts/map-script.js [hunk 2]
```diff
@@ -58,27 +58,41 @@ function initMap() {
...
-  // Load country data after finished loading in geojson.
-  map.data.loadGeoJson('countries.geojson', null, function () {
-    setNewTrend();
-  });
+  map.data.loadGeoJson('countries.geojson', null);
+}
+
+/** 
+ * Gets the selected mode (sentiment or popularity) from the webpage and loads
+ * data correspondingly.
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/map-script.js:66]:
```suggestion
 * corresponding data.
```

File: search-sentiments-map/public/scripts/map-script.js [hunk 3]
```diff
@@ -58,27 +58,41 @@ function initMap() {
...
-    const countryCode = row.getId();
-    let topicData = getCurrentCustomSearchData();
-    let countryData = topicData.countries
-      .filter(countries => countries.country === countryCode);
+/** 
+ * Loads the sentiments or search interests for all countries from Datastore. 
+ * @param {boolean=} sentimentMode Whether the result to obtain is the sentiments 
+ * (search interest otherwise).
+ */
+function loadCountryData(sentimentMode=true) {
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/map-script.js:85]: For boolean params we use the format `isEnabled` or 'shouldUseX`.

So this would be `isSentimentMode` or `shouldLoadSentimentData`

Comment (hayleejane3) [inline search-sentiments-map/public/scripts/map-script.js:85]:
```suggestion
function loadCountryData(sentimentMode = true) {
```

File: search-sentiments-map/public/scripts/map-script.js [hunk 4]
```diff
@@ -58,27 +58,41 @@ function initMap() {
...
 
-/** Loads the country sentiment score from Datastore. */
-function loadCountryData() {
-  map.data.forEach(function (row) {
-    const countryCode = row.getId();
-    let topicData = getCurrentCustomSearchData();
-    let countryData = topicData.countries
-      .filter(countries => countries.country === countryCode);
+/** 
+ * Loads the sentiments or search interests for all countries from Datastore.
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/map-script.js:81]:
```suggestion
 * Loads the sentiments or search interests for all countries from Datastore depending on whether or not the mode has been specified. 
```

File: search-sentiments-map/public/scripts/map-script.js [hunk 5]
```diff
@@ -58,27 +58,41 @@ function initMap() {
...
-function loadCountryData() {
-  map.data.forEach(function (row) {
-    const countryCode = row.getId();
-    let topicData = getCurrentCustomSearchData();
-    let countryData = topicData.countries
-      .filter(countries => countries.country === countryCode);
+/** 
+ * Loads the sentiments or search interests for all countries from Datastore. 
+ * @param {boolean=} sentimentMode Whether the result to obtain is the sentiments 
+ * (search interest otherwise).
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/map-script.js:83]:
```suggestion
 * @param {boolean=} sentimentMode Whether the result to obtain is the sentiment data. Search Interest data will be loaded otherwise.
```

File: search-sentiments-map/public/scripts/modal-script.js [hunk 6]
```diff
@@ -22,7 +22,7 @@ function onClickCountry(e) {
   // Update Modal with information for relevant country.
   const countryName = e.feature.getProperty('name');
   const countryId = e.feature.getId();
-  const countryData = e.feature.getProperty('country_data').toLocaleString();
+  //const countryData= e.feature.getProperty('country_data').toLocaleString();
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/modal-script.js:25]: intentional? If so, add a todo/comment. If not needed now and will not be needed, delete.

File: search-sentiments-map/public/scripts/modal-script.js [hunk 7]
```diff
@@ -55,25 +55,25 @@ function setCountryTrends(countryCode) {
...
   } else {
-    let results = countryData[0].results;
-
+    resultElement.innerHTML += 'Popularity Score: ' + 
+        countryData[0].interest + '<br>';
     resultElement.innerHTML += 'Average Sentiment Score: ' + 
         countryData[0].averageSentiment + '<br>';
-
+    
+    // Get search results of the specified country.
```
Comment (hayleejane3) [inline search-sentiments-map/public/scripts/modal-script.js:75]:
```suggestion
    // Get search results for the specified country.
```

File: search-sentiments-map/public/style.css [hunk 8]
```diff
@@ -68,4 +68,43 @@
...
+  left: 4px;
+  bottom: 4px;
+  background-color: white;
+  -webkit-transition: .4s;
+  transition: .4s;
+}
+
+input:checked + .slider:before {
+  -webkit-transform: translateX(46px);
+  -ms-transform: translateX(46px);
```
Comment (hayleejane3) [inline search-sentiments-map/public/style.css:108]: Alphabetical order.

File: search-sentiments-map/routes/search.js [hunk 9]
```diff
@@ -106,30 +100,31 @@ async function retrieveGlobalTrends() {
  */
 async function updateSearchResultsForTopic(query) {
   let countriesData = [];
+  await searchInterestsModule.getGlobalSearchInterests(query)
+      .then(async (searchInterests) => {
+    // When testing, make i < 3 countries.
```
Comment (hayleejane3) [inline search-sentiments-map/routes/search.js:105]:
```suggestion
    // Note: Use i < 3 countries when testing.
```

File: search-sentiments-map/routes/search.js [hunk 10]
```diff
@@ -106,30 +100,31 @@ async function retrieveGlobalTrends() {
...
-  // When testing, make i < 3 countries.
-  for (let i = 0; i < json.length; i++) {
-    // Use 100 queries per minute limit for the Custom Search API, and include
-    // a pause of 1 minute to prevent surpassing limit.
-    if (i !== 0 && i % 100 === 0) {
-      await new Promise(resolve => setTimeout(resolve, 60000));
+      // Use 100 queries per minute limit for the Custom Search API, and include
+      // a pause of 1 minute to prevent surpassing limit.
+      if (i !== 0 && i % 100 === 0) {
+        await new Promise(resolve => setTimeout(resolve, 60000));
```
Comment (hayleejane3) [inline search-sentiments-map/routes/search.js:115]: Constants for these numbers.

File: search-sentiments-map/routes/search-interests.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+ * Calls the API to get the popularity of the given topic in countries across
+ * the world during the past 7 days. Values are on a scale from 0 to 100,
+ * where 100 is the location with the most popularity.
+ * @return {Array<JSON>} An array of popularity values for each country.
+ */
+async function getGlobalSearchInterests(topic) {
+  let interestByRegion;
+  await googleTrends.interestByRegion({
+    keyword: topic,
+    startTime: new Date(Date.now() - 7 * 24 * 60 * 60000),
```
Comment (carmenBen) [inline search-sentiments-map/routes/search-interests.js:27]: Add a constant for this time.

File: search-sentiments-map/public/scripts/modal-script.js [hunk 12]
```diff
@@ -55,25 +55,25 @@ function setCountryTrends(countryCode) {
  * @param {string} countryCode Two letter country code for selected country.
  */
 function displayTopResultsForCurrentTrend(countryCode) {
-  let topic = getCurrentTrend;
-  let topicData = getCurrentCustomSearchData();
-  let date = new Date(topicData.timestamp);
-  let resultElement = document.getElementById('search-results-tab');
+  let dataByCountry = getCurrentCustomSearchData().dataByCountry;
+  let date = new Date(getCurrentCustomSearchData().timestamp);
+  let resultElement =  document.getElementById('search-results-tab');
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/modal-script.js:60]: Delete extra space.

File: search-sentiments-map/public/scripts/current-data-script.js [hunk 13]
```diff
@@ -35,19 +35,13 @@ function getCurrentCustomSearchData() {
...
-      'Worldwide sentiments of search results for "' + currentTrend + '"';
+  currentTrend = trend;
 
-  // Reload map with new sentiment data and relevant coloring.
-  fetch('/search/' + currentTrend)
-      .then(resultsJsonArray => resultsJsonArray.json()).then(topicResults => {
-        currentCustomSearchData = topicResults;
-        loadCountryData();
-      });
+  fetch('/search/' + trend)
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/current-data-script.js:40]: Should use currentTrend and not trend.

File: search-sentiments-map/public/scripts/map-script.js [hunk 14]
```diff
@@ -58,27 +58,41 @@ function initMap() {
...
+  map.data.loadGeoJson('countries.geojson', null);
+}
+
+/** 
+ * Gets the selected mode (sentiment or popularity) from the webpage and loads
+ * data correspondingly.
+ */
+function loadCountryDataByMode() {
+  let sentimentMode = !document.getElementById('sentiment-popularity-check').checked;
+  console.log('sentiment mode', sentimentMode);
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/map-script.js:70]: Can we delete this console.log?

File: search-sentiments-map/public/scripts/map-script.js [hunk 15]
```diff
@@ -58,27 +58,41 @@ function initMap() {
...
+    let dataByCountry = getCurrentCustomSearchData().dataByCountry;
+    let countryData = dataByCountry.filter(data => data.country === row.getId());
+
     let dataVariable = 0;
-    if (countryData.length != 0) {
-      dataVariable = countryData[0].averageSentiment;
-      // TODO(ntarn): Remove console.log statements when finished debugging.
-      console.log('ntarn debug loadCountryData: ' + 'country' +
-          countryData[0].country + ' averageSentiment: ' + dataVariable);
+    if (countryData.length == 0) {
```
Comment (carmenBen) [inline search-sentiments-map/public/scripts/map-script.js:91]: Change to ===.

File: search-sentiments-map/public/scripts/map-script.js (hunk 3 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/map-script.js:85]: got it

File: search-sentiments-map/public/scripts/map-script.js (hunk 4 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/map-script.js:81]: I'm using "depending on what mode has been specified"

File: search-sentiments-map/public/scripts/map-script.js (hunk 5 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/map-script.js:83]: ok

File: search-sentiments-map/public/scripts/modal-script.js (hunk 6 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/modal-script.js:25]: Deleting it

File: search-sentiments-map/public/style.css (hunk 8 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/style.css:108]: oh sorry I forgot

File: search-sentiments-map/routes/search.js (hunk 10 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/routes/search.js:115]: ok

File: search-sentiments-map/public/scripts/current-data-script.js (hunk 13 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/current-data-script.js:40]: Wouldn't those two be the same tho?

File: search-sentiments-map/public/scripts/modal-script.js (hunk 12 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/modal-script.js:59]: Did it but not sure if that's what you mean

File: search-sentiments-map/public/scripts/map-script.js (hunk 14 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/map-script.js]: yup

File: search-sentiments-map/index.yaml [hunk 16]
```diff
@@ -1,6 +1,6 @@
 indexes:
 
-- kind: CustomSearchTopic
+- kind: WorldDataByTopic
```
Comment (ntarn) [inline search-sentiments-map/index.yaml:3]: Why did you change this? I thought Carmen was using it for the custom search results data

File: search-sentiments-map/public/scripts/current-data-script.js [hunk 17]
```diff
@@ -35,19 +35,14 @@ function getCurrentCustomSearchData() {
...
+      .then(resultsJsonArray => resultsJsonArray.json()).then(topicData => {
+    currentCustomSearchData = topicData;
+  }).then(() => {
```
Comment (ntarn) [inline search-sentiments-map/public/scripts/current-data-script.js:43]: Indentation? maybe github is formatting it weirdly

File: search-sentiments-map/public/scripts/map-script.js [hunk 18]
```diff
@@ -58,27 +58,41 @@ function initMap() {
...
+ * Gets the selected mode (sentiment or popularity) from the webpage and loads
+ * corresponding data.
+ */
+function loadCountryDataByMode() {
+  let isSentimentMode = !document.getElementById('sentiment-popularity-check').checked;
+
+  const topicHeader = document.getElementById('topic-header');
+  topicHeader.innerText = isSentimentMode ?
+      'Worldwide sentiments of search results for "' + getCurrentTrend() + '"' :
+      'Worldwide interests for "' + getCurrentTrend() + '"' ;
```
Comment (ntarn) [inline search-sentiments-map/public/scripts/map-script.js:74]:
```suggestion
      'Worldwide interest for "' + getCurrentTrend() + '"' ;
```

File: search-sentiments-map/routes/search.js [hunk 19]
```diff
@@ -239,22 +238,41 @@ async function deleteAncientResults() {
  * @param {string} topic The seach topic the search results are for.
```
Comment (ntarn) [inline search-sentiments-map/routes/search.js:238]:
```suggestion
 * @param {string} topic The search topic the search results are for.
```

File: search-sentiments-map/routes/search-interests.js [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+// License for the specific language governing permissions and limitations under
+// the License.
+
+const googleTrends = require('google-trends-api');
+const POPULARITY_TIMERANGE_7_DAYS_MS = 7 * 24 * 60 * 60000;
+
+/** 
+ * Calls the API to get the popularity of the given topic in countries across
+ * the world during the past 7 days. Values are on a scale from 0 to 100,
+ * where 100 is the location with the most popularity.
```
Comment (ntarn) [inline search-sentiments-map/routes/search-interests.js:21]: add @param

File: search-sentiments-map/public/scripts/current-data-script.js (hunk 13 above)
Comment (carmenBen) [reply search-sentiments-map/public/scripts/current-data-script.js:40]: Yes but its better to use the currentTrend I would think since that is the general variable for it.

File: search-sentiments-map/index.yaml (hunk 16 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/index.yaml:3]: Because we are changing the entity name

File: search-sentiments-map/public/scripts/current-data-script.js (hunk 13 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/current-data-script.js:40]: Since it's in the function where trend is a parameter, I'd say it's better to use trend

File: search-sentiments-map/public/scripts/current-data-script.js (hunk 17 above)
Comment (ChenyuHeidiZhang) [reply search-sentiments-map/public/scripts/current-data-script.js:43]: I'm not sure how to indent here honestly
